app/models/manageiq/providers/azure/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,403 @@
+    persister.networks.build(
+      :hardware    => persister_hardware,
+      :description => description,
+      :ipaddress   => ip_address,
+      :hostname    => hostname,
+    )
+  end
+
+  def hardware_disks(persister_hardware, instance)
+    data_disks = instance.properties.storage_profile.data_disks
+    data_disks.each do |disk|
+      add_instance_disk(persister_hardware, instance, disk)
+    end
+
+    disk = instance.properties.storage_profile.os_disk
+    add_instance_disk(persister_hardware, instance, disk)
+  end
+
+  # Redefine the inherited method for our purposes
+  def add_instance_disk(persister_hardware, instance, disk)
+    if instance.managed_disk?
+      disk_type     = 'managed'
+      disk_location = disk.managed_disk.id
+      managed_disk  = collector.managed_disks.find { |d| d.id.casecmp(disk_location).zero? }
+
+      if managed_disk
+        disk_size = managed_disk.properties.disk_size_gb.gigabytes
+        mode      = managed_disk.sku.name
+      else
+        _log.warn("Unable to find disk information for #{instance.name}/#{instance.resource_group}")
+        disk_size = nil
+        mode      = nil
+      end
+    else
+      disk_type     = 'unmanaged'
+      disk_location = disk.try(:vhd).try(:uri)
+      disk_size     = disk.try(:disk_size_gb).try(:gigabytes)
+
+      if disk_location
+        uri = Addressable::URI.parse(disk_location)
+        storage_name = uri.host.split('.').first
+        container_name = File.dirname(uri.path)
+        blob_name = uri.basename
+
+        storage_acct = collector.storage_accounts.find { |s| s.name.casecmp(storage_name).zero? }
+        mode = storage_acct.sku.name
+
+        if collector.options.get_unmanaged_disk_space && disk_size.nil?
+          storage_keys = collector.account_keys(storage_acct)
+          storage_key  = storage_keys['key1'] || storage_keys['key2']
+          blob_props   = storage_acct.blob_properties(container_name, blob_name, storage_key)
+          disk_size    = blob_props.content_length.to_i
+        end
+      end
+    end
+
+    persister.disks.build(
+      :hardware        => persister_hardware,
+      :device_type     => 'disk',
+      :controller_type => 'azure',
+      :device_name     => disk.name,
+      :location        => disk_location,
+      :size            => disk_size,
+      :disk_type       => disk_type,
+      :mode            => mode
+    )
+  end
+
+  def stacks
+    collector.stacks.each do |deployment|
+      name = deployment.name
+      uid  = deployment.id
+
+      persister_orchestration_stack = persister.orchestration_stacks.build(
+        :ems_ref                => uid,
+        :name                   => name,
+        :description            => name,
+        :status                 => deployment.properties.provisioning_state,
+        :resource_group         => deployment.resource_group,
+        :orchestration_template => persister.orchestration_templates.lazy_find(uid),
```

Comment:
@bzwei so, the easiest way is to no change orchestration_template association, when it's missing, right? Since stack should not even have blank template

Reply:
A stack can have no template associated. but with target refresh I guess if you don't have `:orchestration_template => nil` the existing association remains. In this case the existing association was created by our code during provisioning, not through refresh.
Anyway I agree that we no longer need to load orchestration_template from VMDB.

Reply:
ok, I will move the association elsewhere, only setting it when it's there

For the orchestration_templates themselves, it looks like we do not delete template at all https://github.com/Ladas/manageiq/blob/f77c75b4659f97be95f8976ef58abe521085c963/app/models/orchestration_template.rb#L37 , so we can just send subset everytime (e.g. just fetching template if stack timestamp changed)

so, this could optimize 2nd refresh a lot

Reply:
I'll do this in another PR, possibly extracting it into setting.

Reply:
`OrchestrationStack` has `updated_at` timestamp which is the timestamp of the AR record. Raw template deployment has attribute `deployment.properties.timestamp` which is [The timestamp of the template deployment](https://docs.microsoft.com/en-us/rest/api/resources/deployments/get#deploymentpropertiesextended) according to MS document. We may skip the template update if `updated_at` is newer than `timestamp`.

Reply:
yeah, I was looking at that. I'll do it as last PR (after the targeted refresh) as a configurable optimization in settings (since it will mess up most of the refresh specs, by changing API queries in the 2nd refresh)